src/tools/web/CMakeLists.txt

```diff
@@ -36,15 +36,7 @@ else ()
 		install (CODE "message (\"-- Installing: elektra-web > elektrad\")")
 		install (
 			CODE
-				"execute_process (COMMAND npm install --unsafe-perm WORKING_DIRECTORY ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad OUTPUT_QUIET)"
-			)
-		install (
-			CODE
-				"execute_process (COMMAND npm run build --unsafe-perm WORKING_DIRECTORY ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad OUTPUT_QUIET)"
-			)
-		install (
-			CODE
-				"execute_process (COMMAND npm prune --production --unsafe-perm WORKING_DIRECTORY ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad OUTPUT_QUIET)"
+				"execute_process (COMMAND go build WORKING_DIRECTORY -o ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad)"
```

Comment:
It would be better if we only install here (and already build during the build process).

Reply:
build and install both both compile. The only difference is that `install` puts the executable in $GOPATH/bin

Reply:
The cmake integration is not done yet - it's more trial and error until I get it to work somehow.

Reply:
@markus2330 I'm really fighting with CMake - maybe you can help me. See [this](https://gist.github.com/raphi011/a605e8eac19e040558932c957e745d95) gist. 

I  temporarily copied the contents from the base/Dockerfile to the elektrad/Dockerfile so I could install go 1.13 (afaik it's not possible to reference local / relative base Dockerfiles).

I want cmake to `go build` the elektrad package and put the resulting binary into `/usr/local/share/elektra/tool_data/web/elektrad` (within the image). So I can run it run it with `./elektrad`. 

Any ideas?

Reply:
The CMake command 

```cmake
execute_process (COMMAND go build WORKING_DIRECTORY -o ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad)
```

seems to be broken. Currently the code supplies (a list of) two arguments:

- `-o`, and
- `${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad`

to the argument `WORKING_DIRECTORY` of the [`execute_process` CMake command](https://cmake.org/cmake/help/v3.15/command/execute_process.html). 

> I want cmake to `go build` the elektrad package and put the resulting binary into `/usr/local/share/elektra/tool_data/web/elektrad` (within the image). So I can run it run it with `./elektrad`. 

Do you care about the working directory of `go build`? If not, then something like

```cmake
execute_process (COMMAND go build -o "${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad")
```

should do the job. If you want to use the directory of the CMake file as current working directory, then

```cmake
execute_process (COMMAND go build -o "${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad" WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}")
```

should work. Please note, that I did **not** test any of the two commands above.

Reply:
Thanks for the quick reply - good catch `WORKING_DIRECTORY -o ...` is an obvious mistake, although this line has changed a few dozen times already trying to make it work. I'll try your suggestions this evening.

Reply:
@sanssecours still did't work. I removed the line that copies the whole elektrad directory to the ${install_directory} and [tried to](https://gist.github.com/raphi011/a605e8eac19e040558932c957e745d95#file-src-tools-web-cmakelists-txt) set the output of the go build command (`COMMAND go build -o ${CMAKE_INSTALL_PREFIX}/${install_directory}/elektrad WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/elektrad`) there instead - doesn't work. The folder does not contain the binary.

Reply:
Did you check the output of the command? You can use the argument `OUTPUT_VARIABLE` and `ERROR_VARIABLE` of the [`execute_process`](https://cmake.org/cmake/help/v3.15/command/execute_process.html) command to store the output and then print it using the [`message`](http://cmake.org/cmake/help/v3.15/command/message.html) command.

Reply:
Is that necessary? I've noticed that when I forget to prepend GO111MODULE=on to the command I already a get an error message without adding the OUTPU_VARIABLE / ERROR_VARIABLE arguments.

What's weird though is that when the command appears to work (no output) it is printed in red. 

Going to try the arguments and report back.

Reply:
> Is that necessary?

It might not be, since I do not know if `execute_process` prints the output of commands by default. However, explicitly capturing the output and printing it should not make the debugging situation worse. If you explicitly prefix the output (e.g. using an emoji like `😱`) you can also be sure that `message` actually reports the output of the command `go build` and not some other command.

Reply:
@sanssecours would you or someone else be willing to help me via a screenshare session sometime this week? I'm wasting so much time on this issue and I'm pretty it's a one or twoliner that will fix this

Reply:
> @sanssecours would you or someone else be willing to help me via a screenshare session sometime this week?

I am not a big fan of videoconferencing, but if you want we can just meet (pretty much any time between the week somewhere in Vienna). If that is okay for you, just send me an email and we can fix a date to debug this problem together.

> I'm wasting so much time on this issue and I'm pretty it's a one or twoliner that will fix this

I can also debug this stuff on my own if that helps. For that however, I would need the steps to reproduce the issue.

Reply:
That would be awesome. How about friday ~1700?

Reply:
> How about friday ~1700?

Sounds good. Where do you want to meet? I would recommend somewhere at one of Vienna’s  universities (maybe in the [“Lernraum” in the first floor of the “Freihaus”](https://www.fsinf.at/lernen#Ehemaliger_Speisesaal_der_Mensa)), since then we do not need to worry about internet access.

Reply:
Perfect!

Reply:
Until Friday then. By the way, I also just sent you my mobile phone per email, in case we do not find each other immediately, or there are other unforeseen problems with our meeting.

Reply:
Just a quick update: I removed and change some code in a [debug commit](https://github.com/sanssecours/elektra/commit/41134577ff350d2d92e0cdb831d9f8b53c4a0bdb). If I use the command:

```sh
docker build -t elektra-webui -f scripts/docker/webui/base/Dockerfile .
```

in the root directory of the repository, then `cmake` prints the following error message:

```
-- Installing: elektra-web > elektrad
find_handler.go:7:2: cannot find package "github.com/ElektraInitiative/go-elektra/kdb" in any of:
	/usr/lib/go-1.10/src/github.com/ElektraInitiative/go-elektra/kdb (from $GOROOT)
	/root/go/src/github.com/ElektraInitiative/go-elektra/kdb (from $GOPATH)
main.go:7:2: cannot find package "github.com/gorilla/mux" in any of:
	/usr/lib/go-1.10/src/github.com/gorilla/mux (from $GOROOT)
	/root/go/src/github.com/gorilla/mux (from $GOPATH)
```

.

Reply:
I took another look at the CMake code. The branch [`🌍` of my copy of the repo](https://github.com/sanssecours/elektra/commits/🌍) contains the results. CMake now builds `elektrad` with `go build` in the build phase and install the binary in the install phase. The code still needs some work, but the basic functionality should be fine. At least on my machine everything works as expected (, if I install Elektra, Go and [Ninja](https://ninja-build.org) via Homebrew):

```sh
# The following commands assume that you are in the root of the repository
# in the branch `🌍`
mkdir build
cd build
cmake .. -GNinja -DTOOLS="kdb;web"
ninja
# After the step above `elektrad` exists in `build/src/tools/web/elektrad`
ls src/tools/web/elektrad/
#> elektrad
ninja install
# Now a copy of `elektrad` should be in the directory `/usr/local/share/elektra/tool_data/web/`
ls /usr/local/share/elektra/tool_data/web/
#> elektrad
```

.

Reply:
Hi, didn't work for me :( 

```
[raphi@localhost build]$ git branch
  master
* 🌍
[raphi@localhost build]$ ls src/tools/web/
CMakeFiles  cmake_install.cmake  CTestTestfile.cmake
```

Reply:
> Hi, followed your instructions but it didn't work for me :(

I assume somehow your setup is a little bit different. Did you maybe forget to install [YAJL](http://lloyd.github.io/yajl/)? I rewrote the [Cirrus configuration file in my latest commit](https://github.com/sanssecours/elektra/commit/01a1e4d67e75cd64bd10407cdaa4426fb66fd83c). It now more or less uses the steps given above to build and install Elektra. According to the [output of Cirrus](https://cirrus-ci.com/task/4859587661922304?command=main#L291-L303) everything seems to work as expected.

Reply:
yajl is installed

Reply:
Can you please add the output of CMake and Ninja on your machine below? Maybe that will help me to decipher what has gone wrong.

Reply:
I ran the exact same commands on my laptop - and now it miraculously works.. Apparently something is wrong with my setup at home. Will have to check the logs, can post them here in the evening. 

Thanks for your support @sanssecours !

Reply:
When doing modifications with cmake it is best to completely remove the build directory before investigating any strange error further.

Reply:
Deleted it before every attempt

Reply:
> I assume somehow your setup is a little bit different. Did you maybe forget to install [YAJL](http://lloyd.github.io/yajl/)?

Looking closely at the logs revealed that yajl was installed, but the wrong version. You were completely right! It's working at home now too.

Reply:
Next problem: If I try to build the elektrad dockerfile it fails because elektrad already expects the elektra shared libraries (elektra.h, kdb.h) to be installed and available, which is not the case. 

My idea would be to simultaneously build & install in the install step to circumvent this..

Reply:
I think it would be great, if you use the libraries from the current Elektra build. Then you do not need an installed version of Elektra. In the CMake code you could add the Elektra libraries as dependencies for `go build`. Then `go build` would only be executed after Elektra’s C libraries already exist in the build directory. I do not know how hard it is to use a non-standard include and library directory in `go build` though…

Reply:
you mean copy the kdb.h and elektra.h files into the go-elektra repo? this would need to be updated periodically.

Reply:
oh, now i know what you mean. I don't think this is possible since elektra-go is a seperate repository that uses package-config to find the shared library.

Reply:
> you mean copy the kdb.h and elektra.h files into the go-elektra repo? this would need to be updated periodically.

Nope, I mean that we use the libelektra libraries in the subdirectory `lib` of the build directory to build the Go module. I do not know how much work that is though. We probably also would have to fix the library paths, if the build system installs `elektrad`. All in all, I probably do know too little about the Go module. Why do we need an installed version of Elektra?

Reply:
so assuming your approach does _not_ work - is it possible to build & install at the same time and **after** the library itself is installed?

Reply:
> is it possible to build & install at the same time and after the library itself is installed?

While that might be possible it sounds like a bad idea. I think we should create a [CMake find module](https://gitlab.kitware.com/cmake/community/wikis/doc/tutorials/How-To-Find-Libraries) for Elektra just as [we do for other libraries like YAJL](https://github.com/ElektraInitiative/libelektra/blob/master/scripts/cmake/Modules/FindYajl.cmake). Afterwards we only include the Go module, if CMake detected an installed version of Elektra.

Reply:
But this is never true for the elektrad dockerfile for example, since elektra is built from scratch there?

Reply:
> But this is never true for the elektrad dockerfile for example, since elektra is built from scratch there?

That might be true currently. However, I do not see a reasons as to why we should not be able to either 

- install some version fo Elektra in a Docker image, or 
- build and install Elektra without the Go module 

, and  then build (and probably also install Elektra) with the Go module afterwards. As already mentioned above, just using the build version of Elektra’s header and library files would probably be the most convenient option for someone that wants to use `elektrad` though.

Reply:
> As already mentioned above, just using the build version of Elektra’s header and library files would probably be the most convenient option for someone that wants to use elektrad though.

Yes but I do not see this happening with how the go toolchain works for dependencies with c / cgo.

Reply:
So our current options for the docker image are

- Install elektra from some other source before building elektrad?
- Compiling / Installing elektra twice, once without elektrad tool and once including elektrad

.. right?

Reply:
Those are two options I see. Of course there can be other options I did not think of.

Reply:
Compiling Elektra several times is a very unsatisfactory solution.

What exactly is the problem so that the go module cannot work with a non-installed Elektra (from a source tree). What does the go module need next to the header files?

Reply:
Well `elektra-go` uses pkg-config. I just remembered that you can configure the path it traverses for `.pc` files via an environment variable `PKG_CONFIG_PATH`, this would need to be set to the previously built .pc files pointing to the correct header files.

Reply:
Can you guys help me with 
* making sure elektrad is built _after_ these pc files are created (if this is not the case already)
* getting the correct paths for the .pc files so I'm able to pass it to `PKG_CONFIG_PATH`

Reply:
> Can you guys help me with
> 
> * making sure elektrad is built _after_ these pc files are created (if this is not the case already)

I think for that to work you only have to specify

```cmake
"${CMAKE_BINARY_DIR}/scripts/cmake/elektra.pc"
```

as dependency (argument `DEPENDS`) in `add_custom_command`.

> * getting the correct paths for the .pc files so I'm able to pass it to `PKG_CONFIG_PATH`

You can use `cmake -E env` to specify environments variables for `go build`, as we do in the code:

https://github.com/ElektraInitiative/libelektra/blob/9c73c76ad4c0f616fdba6ff76efe5908fe681a5e/scripts/cmake/Modules/LibAddMacros.cmake#L557-L570

for the command `ronn` (`${RONN_LOC}`). As for the value of `PKG_CONFIG_PATH `, I think

```
${CMAKE_BINARY_DIR}/scripts/cmake
```

should do the job.

Reply:
I am afraid this alone will not help as you do not want `.pc` files which point to the installed header files but instead a `.pc` file which points to the headers in the source&build directory.

If there is really no other way to tell go to use other header files, it is best to create your own `.pc` file exactly for this purpose. The `.pc` format is quite simple.

For the other problems, @sanssecours already described solutions.